src/term/parser/parser.rs

```diff
@@ -205,47 +212,100 @@ where
   choice((pat_nam, pat_tup))
 }
 
-fn definition<'a, I>() -> impl Parser<'a, I, (Name, Definition), extra::Err<Rich<'a, Token>>>
+fn rule_pat<'a, I>() -> impl Parser<'a, I, RulePat, extra::Err<Rich<'a, Token>>>
+where
+  I: ValueInput<'a, Token = Token, Span = SimpleSpan>,
+{
+  recursive(|rule_pat| {
+    let var = name().map(RulePat::Var).boxed();
+
+    let ctr = name()
+      .then(rule_pat.clone().repeated().collect())
+      .map(|(nam, xs)| RulePat::Ctr(nam, xs))
+      .delimited_by(just(Token::LParen), just(Token::RParen))
+      .boxed();
+
+    choice((var, ctr))
+  })
+}
+
+fn rule<'a, I>() -> impl Parser<'a, I, (Name, Rule), extra::Err<Rich<'a, Token>>>
 where
   I: ValueInput<'a, Token = Token, Span = SimpleSpan>,
 {
-  let lhs = choice((name(), name().delimited_by(just(Token::LParen), just(Token::RParen))));
+  let def_name =
+    name().try_map(|Name(nam), span| if nam != "data" { Ok(Name(nam)) } else { Err(Rich::custom(span, "")) });
```

Comment:
Do we need to check that it's not called `data`?
If it's not inside `()` it'll be parsed as a datatype declaration and we don't need to check it here.
If it's inside `()` then we are certain that this is the name of function and there is no ambiguity, although it might not be best practice to choose this name.